common/rpc/interceptor/namespace_handover.go

```diff
@@ -0,0 +1,144 @@
+)
+
+const (
+	ctxTailRoom = time.Millisecond * 100
+)
+
+var _ grpc.UnaryServerInterceptor = (*NamespaceHandoverInterceptor)(nil).Intercept
+
+type (
+	// NamespaceHandoverInterceptor handles the namespace in handover replication state
+	NamespaceHandoverInterceptor struct {
+		namespaceRegistry namespace.Registry
+		timeSource        clock.TimeSource
+		enabledForNS      dynamicconfig.BoolPropertyFnWithNamespaceFilter
+		metricsHandler    metrics.Handler
+		logger            log.Logger
+	}
+)
+
+func NewNamespaceHandoverInterceptor(
+	enabledForNS dynamicconfig.BoolPropertyFnWithNamespaceFilter,
+	namespaceRegistry namespace.Registry,
+	metricsHandler metrics.Handler,
+	logger log.Logger,
+	timeSource clock.TimeSource,
+) *NamespaceHandoverInterceptor {
+
+	return &NamespaceHandoverInterceptor{
+		enabledForNS:      enabledForNS,
+		namespaceRegistry: namespaceRegistry,
+		metricsHandler:    metricsHandler,
+		logger:            logger,
+		timeSource:        timeSource,
+	}
+}
+
+func (i *NamespaceHandoverInterceptor) Intercept(
+	ctx context.Context,
+	req any,
+	info *grpc.UnaryServerInfo,
+	handler grpc.UnaryHandler,
+) (_ any, retError error) {
+	defer log.CapturePanic(i.logger, &retError)
+
+	if !strings.HasPrefix(info.FullMethod, api.WorkflowServicePrefix) {
+		return handler(ctx, req)
+	}
+
+	// review which method is allowed
+	methodName := api.MethodName(info.FullMethod)
+	namespaceName, err := GetNamespaceName(i.namespaceRegistry, req)
+	if err != nil {
+		return nil, err
+	}
+
+	if i.enabledForNS(namespaceName.String()) {
+		startTime := i.timeSource.Now()
+		defer func() {
+			metrics.HandoverWaitLatency.With(i.metricsHandler).Record(time.Since(startTime))
+		}()
+		_, err = i.waitNamespaceHandoverUpdate(ctx, namespaceName, methodName)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return handler(ctx, req)
+}
+
+func (i *NamespaceHandoverInterceptor) waitNamespaceHandoverUpdate(
+	ctx context.Context,
+	namespaceName namespace.Name,
+	methodName string,
+) (waitTime time.Duration, retErr error) {
+	if _, ok := allowedMethodsDuringHandover[methodName]; ok {
+		return
+	}
+
+	startTime := time.Now()
+	namespaceData, err := i.namespaceRegistry.GetNamespace(namespaceName)
+	if err != nil {
+		return 0, err
+	}
+	if namespaceData.ReplicationState() == enumspb.REPLICATION_STATE_HANDOVER {
+		cbID := uuid.New()
+		waitReplicationStateUpdate := make(chan struct{})
+		i.namespaceRegistry.RegisterStateChangeCallback(cbID, func(ns *namespace.Namespace, deletedFromDb bool) {
+			if ns.ID().String() != namespaceData.ID().String() {
+				return
+			}
+			if ns.State() != enumspb.NAMESPACE_STATE_REGISTERED ||
+				deletedFromDb ||
+				ns.ReplicationState() != enumspb.REPLICATION_STATE_HANDOVER ||
+				!ns.IsGlobalNamespace() {
+				// Stop wait on state change if:
+				// 1. namespace is deleting/deleted
+				// 2. namespace is not in handover
+				// 3. namespace is not global
+				select {
+				case <-waitReplicationStateUpdate:
```

Comment:
what's the purpose of this line? I don't see other senders on the channel.

Reply:
in case the channel is already closed.